File: yahoo_stocks/lib/create_stock_csv.rb
require_relative 'csv_line'
require_relative 'redis_runner'

# /home/ericm/tmp/money/yahooStocks/work.pl
# /cygdrive/c/EKM/yahooStocks 
# "Description","1-Day Price Chg %","Market Cap","P/E","ROE %","Div. Yield %","Debt to Equity","Price to Book","Net Profit Margin (mrq)","Price To Free Cash Flow (mrq)"

def main_sub

  time_var = `date +%Y-%m-%d_%H.%M.%S`;
  some_dir = "./";

  csv_file_list = File.read(ENV["CSV_FILE_LIST"])
  $fileCount = 0;
    
  $dateCommand = "date  +%Y-%m-%d_%H:%M:%S";
  header_string = "Description,1-Day Price Chg \%,Market Cap,P/E,ROE \%,Div. Yield \%,Debt to Equity,Price to Book,Net Profit Margin (mrq),Price To Free Cash Flow (mrq),Sector,Industry";
  command_string = String.new
  csv_file_list.each_line do |line|
  
    puts "Here is line: #{line}"
    command_string = "wget " + line
    # This works - commenting out for now
    # output = `wget #{line}` 
    
    puts "\n-------------\n"
    # sleep( 0.3 )
  end

  @r_runner = RedisRunner.new
  redis_file = ENV["REDIS_FILE"]
  @r_runner.load_keys_from_file(redis_file)

  output_file_path = some_dir + "anotherOutput.txt"
  output_file = File.open(output_file_path, 'w')
  output_file.write(header_string + "\n")
  output_file.close
  files_tmp = Dir.entries(Dir.getwd).sort
  # I could do it in one line, but this seems a bit cleaner to me - enough chaining as it is
  files = files_tmp.map{|f| f if f.include? 'conameu.csv'}.compact
  files.each do |the_file|
    puts "opening the_file: #{the_file}"
    get_industry_and_sector(the_file, output_file_path)
  end
  # $command = "mv -v $outputFile some_dir" . "output.$TIME_VAR.csv";
  # print `$command`;

end # main_sub


def get_industry_and_sector(file_path, output_path )
  csv_array = File.readlines(file_path)
  total_lines = csv_array.length
  puts "Lines in file #{file_path}: #{total_lines}"
  output_file = File.open(output_path, 'a')
  line_num = 0
  industry = String.new
  sector = String.new

  csv_array.each do |next_line|    
    line_num += 1
    csv_line = CsvLine.new(next_line)
    next_line =~ /^(\".*\"),(.*),(.*),(.*),(.*),(.*),(.*),(.*),(.*),(.*)/;
    description = $1
    
    sector_industry = @r_runner.get_value(file_path)
    if (line_num > 3) && (csv_line.valid_line)
      # puts "There it is: #{csv_line.print_line},#{sector_industry} next_line.length: #{next_line.length}"
      output_file.write("#{csv_line.print_line},#{sector_industry}\n")
    end   
  end

  @r_runner.delete_key(file_path)
  output_file.close
end # get_industry_and_sector


puts "Starting !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
main_sub

the_var = ENV["AN_ARM"]
puts "Here is the var: #{the_var}"
the_ed = ENV["EDITOR"]
puts "Here is the ed: #{the_ed}"
redis_file = ENV["REDIS_FILE"]
puts "Here is redis file: #{redis_file}"

# EOF

File: yahoo_stocks/lib/createStockCSV.pl
# /home/ericm/tmp/money/yahooStocks/secondStockCSV.pl
# //cygdrive/c/EKM/yahooStocks 
# "Description","1-Day Price Chg %","Market Cap","P/E","ROE %","Div. Yield %","Debt to Equity","Price to Book","Net Profit Margin (mrq)","Price To Free Cash Flow (mrq)"

mainSub();

# /home/ericm/tmp/money/yahooStocks/work.pl
# //cygdrive/c/EKM/yahooStocks 
# "Description","1-Day Price Chg %","Market Cap","P/E","ROE %","Div. Yield %","Debt to Equity","Price to Book","Net Profit Margin (mrq)","Price To Free Cash Flow (mrq)"


sub mainSub() {
    $TIME_VAR = `date +%Y-%m-%d_%H.%M.%S`;
    $ITME_VAR = chomp $TIME_VAR;
    $some_dir = "./";

    open( CSVFILELIST, "yahoo.stock.csv.files.txt" );
    $fileCount = 0;
    
    $dateCommand = "date  +%Y-%m-%d_%H:%M:%S";
    $headerString = "Description,1-Day Price Chg \%,Market Cap,P/E,ROE \%,Div. Yield \%,Debt to Equity,Price to Book,Net Profit Margin (mrq),Price To Free Cash Flow (mrq),Sector,Industry";

    while ( <CSVFILELIST> ) {
        $fileCount++;
        print `uname -a`;
        print "\n";
        $downloadFileName = $_;
        print `$dateCommand`;
        print "\n";
        $pingCommand = "ping -c4 " + $downloadFileName;
        # print "$downloadFileName\n";
        $output = `wget $downloadFileName`; # | grep -v \'Debt to Equity\' | cat >> output.csv`;
        print "Result for file $downloadFileName: $output\n";
        print "\n-------------\n";
        sleep( 1 );
    } # while ( <CSVFILELIST> )

    close CSVFILELIST;

    # $command = `cat *csv >> zzz.txt`;
    # $command = `rm -v *csv`;
    # $command = `mv -v zzz.txt output.csv`;
    $outputFile = $some_dir . "anotherOutput.txt";
    open( OUTFILE001, ">$outputFile" ) || die "cannot open $outputFile: $!";
    print OUTFILE001 "$headerString\n";
    close OUTFILE001;
    opendir( DIR, $some_dir ) || die "can't opendir $some_dir: $!";
    # ORIG: @dots = grep { /^\./ && -f "$some_dir/$_" } readdir( DIR );
    @dots = grep { /.*conameu.csv$/ && -f "$some_dir/$_" } readdir( DIR ); # just get the *.csv files
    foreach my $dotItem ( @dots ) {
        print "Here is \$dotItem: $dotItem\n";
        getIndustryAndSector( $dotItem, $outputFile );
    } # foreach my $dotItem ( @dots )
    closedir DIR;
    $command = "mv -v $outputFile $some_dir" . "output.$TIME_VAR.csv";
    print `$command`;
    
} # end sub mainSub()

sub getIndustryAndSector(  ) {
    print "starting sub getIndustryAndSector\n";
    # my ( $argImage, $argIncludePath, $argBannerPath ) = @_;
    my ( $csvFileName, $outFile ) = @_;
    my $linesInFile = 0;
    my $downloadFileLine = "";
    # get number of lines in file
    open ( CSVFILE, $csvFileName ) or die "Can't open '$csvFileName': $!";
    $linesInFile++ while ( <CSVFILE> );
    close CSVFILE;
    print "Line count in file $csvFileName: $linesInFile\n"; 
    
    open( CSVFILE002, $csvFileName ) or die "Can't open '$csvFileName': $!"; 
    open( OUTFILE, ">>$outFile" )  or die "Can't open '$outFile': $!";
    my $lineNum = 0;
    my $industry;
    my $sector;
    while ( <CSVFILE002> ) {
        ++$lineNum;
        $downloadFileLine = $_ ;
        # $downloadFileName = chomp $downloadFileName;
        print "Length of line $lineNum: " . length( $downloadFileLine ) . "\n";
        print "Here is the line $lineNum: $downloadFileLine\n";
        if ( $lineNum == $linesInFile ) {
            print "At last line of csv file, skipping\n";
        } else {
            # "Agrium Inc.",0.740,6.44B,6.096,29.042,0.300,"NA",1.585,-3.423,500.449
            # "American Vanguard Corp.",-0.760,282.09M,15.101,12.764,1.000,"NA",1.802,1.566,-9.531
            # # /(\".*\"),(\d+\.\d+),(\d+\.\d+\w)(.*)/;
            $downloadFileLine =~ /^(\".*\"),(.*),(.*),(.*),(.*),(.*),(.*),(.*),(.*),(.*)/;
            
            $Description = $1;
            $OneDayPriceChg = $2;
            $marketCap = $3; # Market Cap
            $PERatio = $4;
            $ROE = $5;
            $DivYield = $6;
            $DebtToEquity = $7;
            $PriceToBook = $8;
            $NetProfitMarginMRQ = $9;
            $PriceToFreeCashFlowMRQ = $10;
            if ( $lineNum == 1 ) {
                print "Keep going\n";
            } elsif ( $lineNum == 2 ) {
                $sector = $Description;
            } elsif ( $lineNum == 3 ) {
                $industry = $Description;
            } else {
                if ( $marketCap =~ /.*B/ ) {
                    print "Market cap is in billions: $marketCap";
                    $marketCap =~ s/B//;
                    $marketCap = $marketCap * 1000000000;
                    print " - new market cap: " . commify( $marketCap ) . "\n";
                } elsif ( $marketCap =~ /.*M/ ) {
                    print "Market cap is in millions: $marketCap";
                    $marketCap =~ s/M//;
                    $marketCap = $marketCap * 1000000;
                    print " - new market cap: " . commify( $marketCap ) . "\n";
                } else {
                    print "Market cap is NA: $marketCap\n";
                } # if ( $marketCap =~ /.*B/ ) 
                
                #########
                if ( $OneDayPriceChg eq "\"NA\"" ) { $OneDayPriceChg = "null"; }
                if ( $marketCap eq "\"NA\"" ) { $marketCap = "null"; } 
                if ( $PERatio eq "\"NA\"" ) { $PERatio = "null"; }
                if ( $ROE eq "\"NA\"" ) { $ROE = "null"; }
                if ( $DivYield eq "\"NA\"" ) { $DivYield = "null"; }
                if ( $DebtToEquity eq "\"NA\"" ) { $DebtToEquity = "null"; }
                if ( $PriceToBook eq "\"NA\"" ) { $PriceToBook = "null"; }
                if ( $NetProfitMarginMRQ eq "\"NA\"" ) { $NetProfitMarginMRQ = "null"; }
                if ( $PriceToFreeCashFlowMRQ eq "\"NA\"" ) { $PriceToFreeCashFlowMRQ = "null"; }
                #########
                
                print OUTFILE "$Description,$OneDayPriceChg,$marketCap,$PERatio,$ROE,$DivYield,$DebtToEquity,$PriceToBook,";
                print OUTFILE "$NetProfitMarginMRQ,$PriceToFreeCashFlowMRQ,$sector,$industry\n";
                
                print "\n";
            } # if ( $lineNum == 1 ) / else / else
        } # if ( $lineNum == $linesInFile )
    } # while ( <CSVFILE002> ) 
    close CSVFILE002;
    close OPTFILE;
    print "ending sub getIndustryAndSector\n";
} # end sub getIndustryAndSector

sub commify {
    my $text = reverse $_[ 0 ];
    $text =~ s/(\d\d\d)(?=\d)(?!\d*\.)/$1,/g;
    return scalar reverse $text;
} # end commify

# EOF

